[TableGen] Add a field to filter out GenericTable entries

wangpc-pp opened the pull request:
A field `FilterClassField` is added to `GenericTable` class, which
is an optional bit field of `FilterClass`. If specified, only those
records with this field being true will have corresponding entries
in the table.

wangpc-pp committed:
[TableGen] Add a field to filter out GenericTable entries

A field `FilterClassField` is added to `GenericTable` class, which
is an optional bit field of `FilterClass`. If specified, only those
records with this field being true will have corresponding entries
in the table.

statham-arm reviewed (changes requested):
The change looks OK to me, but this should surely have a test in `llvm/test/TableGen`.

You don't say what the use case for this is. I'm guessing it might be so you can construct two `GenericTable`s with the same `FilterClass`, so that they select from the same overall set of records, but assign them different `FilterClassField` values so that they include different subsets of the records of that class?

Perhaps demonstrating that in the new test would be a good idea, and maybe mentioning it in the documentation too. (Documentation is at least as much for "what feature can I use to do this thing?" as "what does this feature do?")

michaelmaitland commented on llvm/docs/TableGen/BackEnds.rst:
```diff
@@ -734,22 +738,25 @@ irrelevant.
+  def : AEntry<"Ted",   4, 4, 1>;
+  def : AEntry<"Alice", 4, 5, 1>;
+  def : AEntry<"Costa", 2, 1, 1>;
+  def : AEntry<"Dale",  2, 1, 0>;
```
My concern is that we are generating `AEntry`s with data that isn't needed after the table is built. Once ATable is built, its `IsNeeded` field is no longer needed by the compiler.

I can imagine a scenario where you generate a table of Pseudo instructions and we would need to incur an extra bit per pseudo, per table cost. The reason why it is per table is because if we had two tables we would need `IsNeededForTableA` and `IsNeededForTableB`.

Can we solve this problem without generating "dead" data?

statham-arm replied:
I agree that there should be a way to leave fields of the Tablegen records out of the generated C++. But isn't the `Fields` array in `GenericTable` already a mechanism of this kind? Won't the `IsNeeded` fields be left out of the generated C++ if their names aren't listed in `Fields` for any table citing that record class?

wangpc-pp replied:
Yes! We won't generate unneeded C++ fields because these C++ fields are specified by `Fields`.

wangpc-pp replied:
> @wangpc-pp are you able to make a note of that so it's obvious to anyone who reads the docs?

Yes! Will do!

wangpc-pp commented:
> The change looks OK to me, but this should surely have a test in `llvm/test/TableGen`.

Thanks! I was planning to add a test but I can't find existed tests for `GenericTable`. I will try to add one later.
> 
> You don't say what the use case for this is. I'm guessing it might be so you can construct two `GenericTable`s with the same `FilterClass`, so that they select from the same overall set of records, but assign them different `FilterClassField` values so that they include different subsets of the records of that class?

Yes! This can be a scenario! And I posted another PR #65460 to show the usage.
> 
> Perhaps demonstrating that in the new test would be a good idea, and maybe mentioning it in the documentation too. (Documentation is at least as much for "what feature can I use to do this thing?" as "what does this feature do?")\

Will do, thanks!

wangpc-pp committed:
fixup! [TableGen] Add a field to filter out GenericTable entries

Add a test in `llvm/test/TableGen/generic-tables.td` and make
detailed document

statham-arm reviewed (approved):
LGTM, thanks!

michaelmaitland commented on llvm/docs/TableGen/BackEnds.rst:
```diff
@@ -768,6 +776,7 @@ is guarded by ``GET_ATable_DECL``, while the definitions are guarded by
     { "Ted", 0x4, 0x4 }, // 2
     { "Alice", 0x4, 0x5 }, // 3
     { "Bob", 0x5, 0x3 }, // 4
+    /* { "Dale", 0x2, 0x1 }, // 5 */ // No this line as `IsNeeded` is 0.
```
nit:  ``// Do not generate this line as `IsNeeded` is 0.``

michaelmaitland reviewed (approved):
LGTM

wangpc-pp committed:
fixup! [TableGen] Add a field to filter out GenericTable entries

Address comments
